File: Dart/Structural/facade.dart
/// Complex Classes for a mock video framework
class VideoFile {
  VideoFile({required this.name});

  final String name;
}

abstract interface class CompresionCodec {}

class OggCompresionCodec implements CompresionCodec {}

class MPEG4CompressionCodec implements CompresionCodec {}

class CodecFactory {
  CompresionCodec extract({required VideoFile file}) =>
      file.name.contains('mp4')
          ? MPEG4CompressionCodec()
          : OggCompresionCodec();
}

class BitRateReader {
  VideoFile convert({
    required VideoFile file,
    required CompresionCodec originCodec,
    required CompresionCodec destinationCodec,
  }) {
    print('Converting ${file.name}');
    print(
        'Parsing ${originCodec.runtimeType} to ${destinationCodec.runtimeType}');

    final fileName = file.name
        .replaceAll(originCodec is MPEG4CompressionCodec ? 'mp4' : 'ogg', '');

    final newExtension =
        destinationCodec is MPEG4CompressionCodec ? 'mp4' : 'ogg';

    return VideoFile(name: '$fileName$newExtension');
  }
}

/// End of complex Classes for a mock video framework

/// This is the facade class where we work with the
/// complex video class

/// Simple enum for formats
enum VideoFormat { mpge4, ogg }

/// Facade class to handle a better workflow
class VideoConverter {
  /// Single method convert where we use the framework
  String convert({required String fileName, required VideoFormat format}) {
    final file = VideoFile(name: fileName);
    final sourceCodec = CodecFactory().extract(file: file);

    final destinationCodec = switch (format) {
      VideoFormat.mpge4 => MPEG4CompressionCodec(),
      VideoFormat.ogg => OggCompresionCodec(),
    };

    final convertedFile = BitRateReader().convert(
        file: file,
        originCodec: sourceCodec,
        destinationCodec: destinationCodec);

    return convertedFile.name;
  }
}

void main() {
  /// Instead use the original framwork, we use
  /// our facade class
  final convertor = VideoConverter();
  final videoFile = 'My video.mp4';

  print('Origin: $videoFile');

  final fileResult =
      convertor.convert(fileName: videoFile, format: VideoFormat.ogg);

  print('Result: $fileResult');

  /// Here is the result:
  /*
  Origin: My video.mp4
  Converting My video.mp4
  Parsing MPEG4CompressionCodec to OggCompresionCodec
  Result: My video.ogg
  */
}

File: Swift/Structural/facade.swift
//
//  facade.swift
//
//
//

import Foundation

// MARK: -  Complex Classes for a mock video framework
struct VideoFile {
    let name: String

    init(name: String) {
        self.name = name
    }
}

protocol CompresionCodec {}

struct OggCompresionCodec: CompresionCodec {}

struct MPEG4CompressionCodec: CompresionCodec {}

struct CodecFactory {
    func extract(file: VideoFile) -> any CompresionCodec {
        file.name.contains("mp4") ? MPEG4CompressionCodec() : OggCompresionCodec()
    }
}

struct BitRateReader {
    func convert(
        file: VideoFile,
        originCodec: any CompresionCodec,
        destinationCodec: any CompresionCodec
    ) -> VideoFile {
        print("Converting \(file.name)")
        print("Parsing \(type(of: originCodec)) to \(type(of: destinationCodec))")

        let fileName = file.name.replacingOccurrences(
            of: originCodec is MPEG4CompressionCodec ? "mp4" : "ogg", with: "")

        let newExtension = destinationCodec is MPEG4CompressionCodec ? "mp4" : "ogg"

        return VideoFile(name: "\(fileName)\(newExtension)")
    }
}


// MARK: - This is the facade class where we work with the complex video class

// Simple enum for formats
enum VideoFormat { case mpeg4, ogg }

// Facade class to handle a better workflow
struct VideoConverter {
    // Single method convert where we use the framework
    func convert(fileName: String, format: VideoFormat) -> String {
        let file = VideoFile(name: fileName)
        let sourceCodec = CodecFactory().extract(file: file)

        let destinationCodec: any CompresionCodec = switch format {
        case .mpeg4: MPEG4CompressionCodec()
        case .ogg: OggCompresionCodec()
        }

        let convertedFile = BitRateReader().convert(
            file: file,
            originCodec: sourceCodec,
            destinationCodec: destinationCodec)

        return convertedFile.name
    }
}

// Demo

// Instead of using the original framework, we use
// our facade class
let convertor = VideoConverter()
let videoFile = "My video.mp4"

print("Origin: \(videoFile)")

let fileResult = convertor.convert(fileName: videoFile, format: .ogg)

print("Result: \(fileResult)")

// Here is the result:

// Origin: My video.mp4
// Converting My video.mp4
// Parsing MPEG4CompressionCodec to OggCompresionCodec
// Result: My video.ogg
